libraries/AC_Avoidance/AP_OADijkstra.cpp

```diff
@@ -0,0 +1,550 @@
+        bool complete = false;
+        while (!complete) {
+            // fail if out of space or newest node has invalid distance or from index
+            if ((_path_numpoints >= OA_DIJKSTRA_POLYGON_SHORTPATH_PTS) ||
+                (_short_path_data[nidx].distance_from_idx == OA_DIJKSTRA_POLYGON_SHORTPATH_NOTSET_IDX) ||
+                (_short_path_data[nidx].distance_cm >= OA_DIJKSTRA_POLYGON_VISGRAPH_INFINITY_CM)) {
+                complete = true;
+            } else {
+                // add node's id to path array
+                _path[_path_numpoints] = _short_path_data[nidx].id;
+                _path_numpoints++;
+
+                // we are done if node is the source
+                if (_short_path_data[nidx].id.id_type == AP_OAVisGraph::OATYPE_SOURCE) {
+                    success = true;
+                    complete = true;
+                } else {
+                    // follow node's "distance_from_idx" to previous node on path
+                    nidx = _short_path_data[nidx].distance_from_idx;
+                }
+            }
+        }
+        // update source and destination for by get_shortest_path_point
+        if (success) {
+            _path_source = origin_NE;
+            _path_destination = destination_NE;
+        }
+    }
+
+    return success;
+}
+
+// print out all node data for debug purposes
+void AP_OADijkstra::print_all_nodes()
+{
+    ::printf("Nodes %d -----\n",(int)_short_path_data_numpoints);
+    for (uint8_t i=0; i<_short_path_data_numpoints; i++) {
+        ShortPathNode& curr_node = _short_path_data[i];
+        ::printf("%d: id:%d-%d v:%d dist:%4.1f from:%d\n",
+                (int)i,
+                (int)curr_node.id.id_type,
+                (int)curr_node.id.id_num,
+                (int)curr_node.visited,
+                (double)curr_node.distance_cm,
+                (int)curr_node.distance_from_idx);
+    }
+}
+
+// return point from final path as an offset (in cm) from the ekf origin
+bool AP_OADijkstra::get_shortest_path_point(uint8_t point_num, Vector2f& pos)
+{
+    if ((_path_numpoints == 0) || (point_num >= _path_numpoints)) {
+        return false;
+    }
+
+    // get id from path
+    AP_OAVisGraph::OAItemID id = _path[_path_numpoints - point_num - 1];
+
+    // convert id to a position offset from EKF origin
+    switch (id.id_type) {
+    case AP_OAVisGraph::OATYPE_SOURCE:
+        pos = _path_source;
+        return true;
+    case AP_OAVisGraph::OATYPE_DESTINATION:
+        pos = _path_destination;
+        return true;
+    case AP_OAVisGraph::OATYPE_FENCE_POINT:
+        // sanity check polygon fence has the point
+        if (id.id_num < _polyfence_numpoints) {
+            pos = _polyfence_pts[id.id_num];
+            return true;
+        }
+        return false;
+    }
+
+    // we should never reach here but just in case
+    return false;
+}
+
+// return number of points in shortest path
+uint8_t AP_OADijkstra::shortest_path_numpoints()
```

Comment:
Do we need this helper? It's a one line helper, that isn't used anywhere else and impairs the readability in the one spot it's used (where it's used differently/as part of a debug interface).

Reply:
No we don't.  I originally added it when I thought we needed a semaphore around the final path.  We may need that eventually but for the moment only the backend ever accesses the path so removing. txs